File: frontend/src/backend/scheduledTransactionService.ts
import api from "./api";
import { normalizePageResponse } from "./pageUtils";
import type {
  ScheduledTransactionDto,
  ScheduledTransactionUpdateDto,
  ScheduledTransactionResponseDto,
  ScheduledTransactionPage,
} from "@/types";

export const scheduledTransactionService = {
  getAll(page = 0, size = 20): Promise<ScheduledTransactionPage> {
    return api
      .get("/scheduled-transaction", { params: { page, size } })
      .then((r) => normalizePageResponse(r.data, page, size));
  },

  getById(id: string): Promise<ScheduledTransactionResponseDto> {
    return api.get(`/scheduled-transaction/${id}`).then((r) => r.data);
  },

  create(data: ScheduledTransactionDto): Promise<ScheduledTransactionResponseDto> {
    return api.post("/scheduled-transaction", data).then((r) => r.data);
  },

  update(id: string, data: ScheduledTransactionUpdateDto): Promise<ScheduledTransactionResponseDto> {
    return api.put(`/scheduled-transaction/${id}`, data).then((r) => r.data);
  },

  remove(id: string): Promise<void> {
    return api.delete(`/scheduled-transaction/${id}`);
  },

  firePending(): Promise<{ created: number }> {
    return api.post("/scheduled-transaction/fire").then((r) => r.data);
  },
};

File: backend/src/main/java/Balio/web/model/services/ScheduledTransactionService.java
package Balio.web.model.services;

import Balio.web.enums.TransactionType;
import Balio.web.model.entities.ScheduledTransaction;
import org.springframework.data.domain.Page;

import javax.management.InstanceNotFoundException;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.UUID;

public interface ScheduledTransactionService {

    ScheduledTransaction create(UUID userId, String name, BigDecimal amount, TransactionType type,
                                 UUID accountId, UUID categoryId, boolean affectsBalance,
                                 int freqYears, int freqMonths, int freqWeeks, int freqDays,
                                 LocalDate startDate)
            throws InstanceNotFoundException;

    ScheduledTransaction update(UUID userId, UUID scheduledTxId, String name, BigDecimal amount,
                                 TransactionType type, UUID accountId, UUID categoryId,
                                 Boolean affectsBalance, Integer freqYears, Integer freqMonths,
                                 Integer freqWeeks, Integer freqDays, LocalDate startDate,
                                 Boolean active)
            throws InstanceNotFoundException;

    void delete(UUID userId, UUID scheduledTxId) throws InstanceNotFoundException;

    ScheduledTransaction findById(UUID userId, UUID scheduledTxId) throws InstanceNotFoundException;

    Page<ScheduledTransaction> findAllByUserId(UUID userId, int page, int size);

    /**
     * Fires all pending scheduled transactions for the user.
     * @return the number of transactions created
     */
    int firePending(UUID userId);

    /**
     * Calculate the next execution date for a scheduled transaction.
     */
    LocalDate calculateNextExecution(ScheduledTransaction st);
}
